drivers: added support for Xbee S1 modules

haukepetersen opened the pull request:
~~based on #2666 and #2693.~~
~~NOTE: currently the packet buffer will overflow, this is fixed in #2563.~~

This implementation evolved from @rousselk's Xbee driver (-> #2250). It now comes with full support for the new network stack, implementing `ng_netdev` and using the `ng_pktbuf`.

I put some care into this implementation so it can act as guideline on how to adapt a driver for the 'ng_netdev' interface.

authmillenon commented on drivers/include/xbee.h:
```diff
@@ -0,0 +1,163 @@
+/**
+ * @brief   Default short address used after initialization
+ */
+#define XBEE_DEFAULT_SHORT_ADDR     (0x0230)
```
Can you use the cpuid to set it, if it exists?

haukepetersen replied:
possible, I left it open for future enhancement for now...

haukepetersen replied:
put it in...

authmillenon commented on drivers/xbee/xbee.c:
```diff
@@ -0,0 +1,700 @@
+        return -ENXIO;
+    }
+    /* set device driver */
+    dev->driver = &xbee_driver;
```
Is this possible when #2693 makes driver const?

haukepetersen replied:
yes

haukepetersen commented:
rebased

haukepetersen commented:
rebased again, now all direct dependencies are merged

authmillenon commented on drivers/include/xbee.h:
```diff
@@ -0,0 +1,163 @@
+ * @brief   Default protocol for data that is coming in
+ */
+#ifdef MODULE_NG_SIXLOWPAN
+#define XBEE_DEFAULT_PROTOCOL       NG_NETTYPE_SIXLOWPAN
```
Parenthesis missing

haukepetersen replied:
hm, I never use parens in statements like this...

haukepetersen replied:
but can't hurt either, will fix

authmillenon commented on drivers/include/xbee.h:
```diff
@@ -0,0 +1,163 @@
+#ifdef MODULE_NG_SIXLOWPAN
+#define XBEE_DEFAULT_PROTOCOL       NG_NETTYPE_SIXLOWPAN
+#else
+#define XBEE_DEFAULT_PROTOCOL       NG_NETTYPE_UNDEF
```
Parenthesis missing

authmillenon commented on drivers/include/xbee.h:
```diff
@@ -0,0 +1,163 @@
+ * @param[in]  sleep_pin    GPIO pin that is connected to the SLEEP pin, set to
+ *                          GPIO_NUMOF if not used
+ * @param[in]  status_pin   GPIO pin that is connected to the STATUS pin, set to
+ *                          GPIO_NUMOR if not used
```
`s/NUMOR/NUMOF/`

authmillenon commented on drivers/xbee/xbee.c:
```diff
@@ -0,0 +1,700 @@
+
+    /* set default short address */
+    _set_addr(dev, dev->addr_short, 2);
+    /* load long address (we can not set it, as it is read only) */
```
I would also point out, that this is only true for this particular device, keeping in mind that this is supposed to be a reference implementation

authmillenon commented on drivers/xbee/xbee.c:
```diff
@@ -0,0 +1,700 @@
+    if (size > XBEE_MAX_PAYLOAD_LENGTH) {
+        DEBUG("xbee: Error sending data, payload length exceeds limit\n");
+        ng_pktbuf_release(pkt);
+        return 0;
```
would return `-EBADMSG` or `-ECOMM` or at least something negative (I think I used `-EBADMSG` in `ng_tapnet`).

haukepetersen replied:
ups, had that in before, must have slipped out on the way...

haukepetersen replied:
`-EOVERFLOW` is the correct one when looking into `ng_netdev.h` :-)

authmillenon commented on drivers/xbee/xbee.c:
```diff
@@ -0,0 +1,700 @@
+        dev->tx_buf[3] = API_ID_TX_LONG_ADDR;
+        memcpy(dev->tx_buf + 11, ng_netif_hdr_get_dst_addr(hdr), 8);
+        pos = 13;
+    }
```
Maybe that is documented for xbee somewhere, but how is the source address for the packet selected? For the `netdev` API I introduced `NETCONF_SRC_LEN` for this.

haukepetersen replied:
you can't manually. When sending with a 16 bit destination address, the 16-bit source address is selected automatically and same for the long addresses.

authmillenon commented on drivers/xbee/xbee.c:
```diff
@@ -0,0 +1,700 @@
+            return _get_panid(dev, (uint8_t *)value, max_len);
+        case NETCONF_OPT_PROTO:
+            return _get_proto(dev, (uint8_t *)value, max_len);
+        default:
```
`NETCONF_OPT_ADDR_LEN` and `NETCONF_OPT_SRC_LEN` missing.

haukepetersen replied:
`NETCONF_OPT_SRC_LEN` is not applicable. And for `NETCONF_OPT_ADDR_LEN` I am not sure what to return...

authmillenon replied:
Okay… then I probably have to change my IPHC compression implementation a little :-D

authmillenon commented on drivers/xbee/xbee.c:
```diff
@@ -0,0 +1,700 @@
+        case NETCONF_OPT_NID:
+            return _set_panid(dev, (uint8_t *)value, value_len);
+        case NETCONF_OPT_PROTO:
+            return _set_proto(dev, (uint8_t *)value, value_len);
```
`NETCONF_OPT_SRC_LEN` missing

haukepetersen replied:
not applicable...

authmillenon commented on drivers/xbee/xbee.c:
```diff
@@ -0,0 +1,700 @@
+    }
+    hdr = (ng_netif_hdr_t *)pkt_head->data;
+    hdr->src_l2addr_len = (uint8_t)addr_len;
+    hdr->dst_l2addr_len = (uint8_t)addr_len;
```
Don't know if this is correct. The address length in IEEE 802.15.4 can be of different lengths for source and destination

haukepetersen replied:
not supported by the Xbee device, though.

authmillenon commented on tests/driver_xbee/main.c:
```diff
@@ -0,0 +1,129 @@
+        return -1;
+    }
+    /* register network device */
+    res = ng_netif_add(iface);
```
`nomac` takes care of this.

haukepetersen commented:
addressed comments, added CPUID for default address selection. I will look into the documentation again, if the sending address length can actually be influenced, but I doubt it...

authmillenon commented:
ACK, please squash.

authmillenon commented:
(only merge, when travis is happy ;-))

haukepetersen commented:
squashed, waiting for Travis then.

authmillenon commented:
I think Travis is mopey now… I stopped some outdated builds a few hours ago and now it hangs on one of them -.-

kaspar030 commented:
Architecture wise, a ng_netdev implementation on top of a low-level driver would be vastly preferrable.

haukepetersen commented:
@kaspar030: why introduce another abstraction layer, that does not make sense. The `ng_netdev` is exactly that low-level driver...

kaspar030 commented:
@haukepetersen Sorry, that was unclear.

I'd prefer to have the basic functions that realize communication with the chip seperated (e.g., in a seperate file) from the full netdev implementation.

You definately have functions for initializing the UART, the chip, reading a packet from the chip, sending a packet, getting/setting addresses, handling a ISR, ... which have *nothing* to do with our netdev interface. If they are seperate, it should be possible to write a test application just for that, without dependencies to netdev. And once that functionality is implemented, writing the implementation of the netdev interface should be trivial. But not having everything tied together *will* make things easier in the future.

haukepetersen commented:
Ah, I see. Actually I agree with you - in the case of this driver I just waived this though, as the UART is so simple to program, that it did not make too much sense to separate it...

FrancescoErmini commented:
Hi, Hauke
The test function works fine in my UDOO, maybe you can add this board in the Makefile for export XBEE_UART in UART_1.

I saw that in your library the EE and KY API AT commands (Those for enable  AES encryption)  are missing. As far as I understood from the Datasheet, it's the Xbee Hardware that takes care of encrypt/decrypt  the API frame. 
There are only few things that may  affects your library (see the EE command details in datasheet):
1. The maximum RF packet size is 95 Bytes
2. Receive frames (64 bit only)  use an option bit to indicate that the packet was encrypted
It's not specify which bit, but I think it is one of the reserved bit in the 14th byte of the  64 bit rx frame (see the image below) 
Question: During the _rx_cb, in the fsm mechanism, this byte (14th byte of RX frame) is stored in rx_buf[10], because the first 4 incoming char are not stored, right?
![xbeerxframe](https://cloud.githubusercontent.com/assets/9249292/6845653/073745e2-d3b7-11e4-947a-36942b0f3f67.png)
If you're not planning to implement encryption in the Xbee driver, then I  have to do so for my thesis. The big problem will be combine the API encryption with the 802.15.4 encryption.
 Some problems that may occur..
http://stackoverflow.com/questions/29028157/how-to-decrypt-a-xbee-s1-packet-encrypted-knowing-the-key

haukepetersen commented:
Hi Francesco, 

thanks for testing!

> 1. The maximum RF packet size is 95 Bytes

Hm, it states 100 Byte int he datasheet (see image you posted), so I took that number...


> Question: During the _rx_cb, in the fsm mechanism, this byte (14th byte of RX frame) is stored in rx_buf[10], because the first 4 incoming char are not stored, right?

yes.

Regarding encryption: I did not look into this at all -> the goal was to provide basic radio functionality for now. But feel free to improve the driver and add functionality!

FrancescoErmini commented:
Ok,I'll implement the AES encryption for Xbee.
The 100 byte frame size is correct,only in case of encryption it becomes 95 byte.
I say this because If EE and KY command are set,then the frame size should change. 
![screenshot 2015-03-26 15 21 55](https://cloud.githubusercontent.com/assets/9249292/6848156/e8e23632-d3cb-11e4-9f4d-41997a470e82.png)
Thanks, F

authmillenon commented:
And go \o/
